okhttp/src/main/java/okhttp3/internal/tls/BasicTrustRootIndex.kt

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright (C) 2016 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okhttp3.internal.tls
+
+import java.security.cert.X509Certificate
+import javax.security.auth.x500.X500Principal
+
+/** A simple index that of trusted root certificates that have been loaded into memory.  */
+class BasicTrustRootIndex(vararg caCerts: X509Certificate) : TrustRootIndex {
+  private val subjectToCaCerts = mutableMapOf<X500Principal, MutableSet<X509Certificate>>()
```

Comment:
I think it's worth removing the mutable nature here.  Either be converting to just map, or inlining the whole calculation in functional zealot style :)

Reply:
Made non-mutable without going full functional zealot. The functional zealot part is difficult ’cause we don’t have a multimap type.